File: backend/services/agent_service.py
"""
Service layer for agent operations.

This module handles high-level agent operations that require coordination
between CRUD operations and other services (like agent manager cleanup).
"""

from typing import TYPE_CHECKING

import crud
from sqlalchemy.ext.asyncio import AsyncSession
from utils.helpers import get_pool_key

if TYPE_CHECKING:
    from orchestration import ChatOrchestrator
    from sdk import AgentManager


async def delete_agent_with_cleanup(db: AsyncSession, agent_id: int, agent_manager: "AgentManager") -> bool:
    """
    Delete an agent permanently and cleanup all associated clients.

    Args:
        db: Database session
        agent_id: ID of the agent to delete
        agent_manager: AgentManager instance for cleanup

    Returns:
        True if agent was deleted, False if agent not found
    """
    # Delete from database
    success = await crud.delete_agent(db, agent_id)
    if not success:
        return False

    # Cleanup all clients for this agent across all rooms
    # Use ClientPool's get_keys_for_agent helper
    pool_keys_to_cleanup = agent_manager.client_pool.get_keys_for_agent(agent_id)
    for pool_key in pool_keys_to_cleanup:
        await agent_manager.client_pool.cleanup(pool_key)

    return True


async def remove_agent_from_room_with_cleanup(
    db: AsyncSession, room_id: int, agent_id: int, agent_manager: "AgentManager"
) -> bool:
    """
    Remove an agent from a room and cleanup the associated client.

    Args:
        db: Database session
        room_id: ID of the room
        agent_id: ID of the agent to remove
        agent_manager: AgentManager instance for cleanup

    Returns:
        True if agent was removed, False if room or agent not found
    """
    # Remove from database
    success = await crud.remove_agent_from_room(db, room_id, agent_id)
    if not success:
        return False

    # Cleanup the client for this room-agent pair
    pool_key = get_pool_key(room_id, agent_id)
    await agent_manager.client_pool.cleanup(pool_key)

    return True


async def delete_room_with_cleanup(
    db: AsyncSession, room_id: int, agent_manager: "AgentManager", chat_orchestrator: "ChatOrchestrator" = None
) -> bool:
    """
    Delete a room and cleanup all associated agent clients and orchestrator state.

    Args:
        db: Database session
        room_id: ID of the room to delete
        agent_manager: AgentManager instance for cleanup
        chat_orchestrator: ChatOrchestrator instance for state cleanup (optional for backwards compatibility)

    Returns:
        True if room was deleted, False if room not found
    """
    import logging

    logger = logging.getLogger("AgentService")

    # Get all agents in the room before deletion for cleanup
    agents = await crud.get_agents(db, room_id)

    # Clean up orchestrator state FIRST (interrupts any active processing)
    if chat_orchestrator:
        try:
            logger.info(f"🧹 Cleaning up orchestrator state for room {room_id}")
            await chat_orchestrator.cleanup_room_state(room_id, agent_manager)
        except Exception as e:
            logger.error(f"❌ Error cleaning orchestrator state for room {room_id}: {e}")
            # Continue with deletion even if orchestrator cleanup fails
    else:
        logger.warning(f"⚠️  No chat_orchestrator provided for room {room_id} deletion - state may leak")

    # Delete from database
    success = await crud.delete_room(db, room_id)
    if not success:
        return False

    # Cleanup all clients for this room (with error handling for each client)
    for agent in agents:
        try:
            pool_key = get_pool_key(room_id, agent.id)
            await agent_manager.client_pool.cleanup(pool_key)
            logger.info(f"✅ Cleaned up client for agent {agent.id} in room {room_id}")
        except Exception as e:
            logger.error(f"❌ Error cleaning up client for agent {agent.id} in room {room_id}: {e}")
            # Continue cleaning up other agents even if one fails

    logger.info(f"✅ Room {room_id} deleted successfully")
    return True


async def clear_room_messages_with_cleanup(
    db: AsyncSession, room_id: int, agent_manager: "AgentManager", chat_orchestrator: "ChatOrchestrator" = None
) -> bool:
    """
    Clear all messages from a room and reset agent sessions.

    When messages are cleared, we also need to:
    1. Interrupt any active processing
    2. Clear agent session IDs (so they start fresh conversations)
    3. Cleanup agent clients (so they don't try to resume invalid sessions)

    Args:
        db: Database session
        room_id: ID of the room to clear messages from
        agent_manager: AgentManager instance for cleanup
        chat_orchestrator: ChatOrchestrator instance for interrupting active processing (optional)

    Returns:
        True if messages were cleared, False if room not found
    """
    import logging

    logger = logging.getLogger("AgentService")

    # Get all agents in the room for cleanup
    agents = await crud.get_agents(db, room_id)
    logger.info(f"🗑️  Clearing room {room_id} messages | Agents: {len(agents)}")

    # Interrupt any active processing FIRST (don't save partial responses since we're clearing all)
    if chat_orchestrator:
        try:
            logger.info(f"🛑 Interrupting room {room_id} processing before clearing messages")
            await chat_orchestrator.interrupt_room_processing(room_id, agent_manager, save_partial_responses=False)
        except Exception as e:
            logger.error(f"❌ Error interrupting room {room_id}: {e}")
            # Continue with cleanup even if interruption fails

    # Delete all messages
    success = await crud.delete_room_messages(db, room_id)
    if not success:
        return False
    logger.info(f"✅ Deleted all messages from room {room_id}")

    # Clear all session IDs for this room (fresh start)
    # This is done by deleting the RoomAgentSession records
    from database import serialized_write
    from models import RoomAgentSession
    from sqlalchemy import delete

    async with serialized_write():
        await db.execute(delete(RoomAgentSession).where(RoomAgentSession.room_id == room_id))
        await db.commit()
    logger.info(f"✅ Cleared all session IDs for room {room_id}")

    # Cleanup all clients for this room (they may have stale session references)
    for agent in agents:
        try:
            pool_key = get_pool_key(room_id, agent.id)
            logger.info(f"🧹 Calling client_pool.cleanup for {pool_key}")
            await agent_manager.client_pool.cleanup(pool_key)
        except Exception as e:
            logger.error(f"❌ Error cleaning up client for agent {agent.id}: {e}")
            # Continue with other agents

    # Invalidate caches so subsequent polls don't return stale messages
    crud.invalidate_room_cache(room_id)
    logger.info(f"🧽 Invalidated room {room_id} cache after clearing messages")

    logger.info(f"✅ Room {room_id} cleared successfully")
    return True

File: backend/src/services/agent-service.ts
/**
 * Deletions that reach further than the database: a `crud/` delete plus eviction
 * of the warm Claude sessions it invalidated. Skipping the eviction is a
 * correctness bug — a `SessionPool` entry outlives its row, and the next turn
 * for a *recycled* SQLite rowid gets a subprocess still holding the previous
 * agent's prompt.
 *
 * **Ordering: interrupt, delete, evict** — interrupting later leaves a window
 * where an in-flight turn's `createMessage` hits a dangling foreign key. **A
 * failed eviction never fails the operation,** since the row is already gone and
 * retrying would 404; each is isolated so one wedged subprocess strands nothing.
 */


import { invalidateRoomCache } from '@/crud/cache-invalidation'
import { deleteAgent } from '@/crud/agents'
import { deleteRoomMessages } from '@/crud/messages'
import { deleteRoom, getAgentsInRoom, removeAgentFromRoom } from '@/crud/rooms'
import type { Db } from '@/db'
import { deleteRoomAgentSessions } from '@/crud/sessions'
import { getLogger } from '@/infrastructure/logging/logger'
import { sessionKeyOf } from '@/sdk/client/session'
import type { SessionPool } from '@/sdk/client/session-pool'

const logger = getLogger('AgentService')

/** A seam: depending on the one method avoids a cycle with `RoomOrchestrator`. */
export interface RoomInterrupter {
  interruptRoom(roomId: number): Promise<void>
}

export class AgentService {
  /**
   * @param orchestrator Interrupts an in-flight turn before its room is
   *   dismantled. Optional: omitting it only warns, since a room the user asked
   *   to delete has to go away regardless.
   */
  constructor(
    private readonly pool: SessionPool,
    private readonly orchestrator: RoomInterrupter | null = null,
  ) {}

  /** `false` when it did not exist. Eviction is by *agent*: it may be in many rooms. */
  async deleteAgentWithCleanup(db: Db, agentId: number): Promise<boolean> {
    if (!deleteAgent(db, agentId)) return false

    for (const key of this.pool.keysForAgent(agentId)) {
      await this.evictQuietly(key, `agent ${agentId}`)
    }

    return true
  }

  /** `false` when the membership did not exist. Other rooms stay warm. */
  async removeAgentFromRoomWithCleanup(db: Db, roomId: number, agentId: number): Promise<boolean> {
    if (!removeAgentFromRoom(db, roomId, agentId)) return false

    await this.evictQuietly(sessionKeyOf({ roomId, agentId }), `room ${roomId}`)
    return true
  }

  /**
   * `pool.evictRoom` scans by key prefix, so it needs no pre-read of the
   * membership (which cascades away with the room).
   */
  async deleteRoomWithCleanup(db: Db, roomId: number): Promise<boolean> {
    if (this.orchestrator) {
      try {
        logger.info(`Cleaning up orchestrator state for room ${roomId}`)
        await this.orchestrator.interruptRoom(roomId)
      } catch (error) {
        logger.error(`Error cleaning orchestrator state for room ${roomId}: ${String(error)}`)
      }
    } else {
      logger.warning(
        `No orchestrator provided for room ${roomId} deletion - state may leak`,
      )
    }

    if (!deleteRoom(db, roomId)) return false

    await this.evictRoomQuietly(roomId)

    logger.info(`Room ${roomId} deleted successfully`)
    return true
  }

  /**
   * Wipe a room's transcript and make its agents forget it. Messages alone are
   * not enough: each `room_agent_sessions` row points at a CLI conversation
   * still holding them, which the next turn would `resume`.
   */
  async clearRoomMessagesWithCleanup(db: Db, roomId: number): Promise<boolean> {
    const agents = getAgentsInRoom(db, roomId)
    logger.info(`Clearing room ${roomId} messages | Agents: ${agents.length}`)

    if (this.orchestrator) {
      try {
        logger.info(`Interrupting room ${roomId} processing before clearing messages`)
        await this.orchestrator.interruptRoom(roomId)
      } catch (error) {
        logger.error(`Error interrupting room ${roomId}: ${String(error)}`)
      }
    }

    if (!deleteRoomMessages(db, roomId)) return false
    logger.info(`Deleted all messages from room ${roomId}`)

    deleteRoomAgentSessions(db, roomId)
    logger.info(`Cleared all session IDs for room ${roomId}`)

    await this.evictRoomQuietly(roomId)

    // `deleteRoomMessages` does not sweep its own cache; without this the poll
    // would keep serving the deleted transcript for a TTL.
    invalidateRoomCache(roomId)
    logger.info(`Invalidated room ${roomId} cache after clearing messages`)

    logger.info(`Room ${roomId} cleared successfully`)
    return true
  }

  private async evictQuietly(key: string, context: string): Promise<void> {
    try {
      await this.pool.evict(key)
      logger.info(`Cleaned up session ${key} for ${context}`)
    } catch (error) {
      logger.error(`Error cleaning up session ${key} for ${context}: ${String(error)}`)
    }
  }

  private async evictRoomQuietly(roomId: number): Promise<void> {
    try {
      await this.pool.evictRoom(roomId)
      logger.info(`Cleaned up sessions for room ${roomId}`)
    } catch (error) {
      logger.error(`Error cleaning up sessions for room ${roomId}: ${String(error)}`)
    }
  }
}

